File: PDOO-Napakalaki/Napakalaki/NapakalakiRuby/lib/player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "treasure.rb"
require_relative "treasure_kind.rb"
require_relative "bad_consequence.rb"
require_relative "dice.rb"
require_relative "combat_result.rb"

class Player
  
  @@MAXLEVEL = 10
  
  attr_accessor :name, :level, :death, :canISteal, :visibleTreasures, :hiddenTreasures, :enemy, :pendingBadConsequence, :MAXLEVEL
  
  def initialize(name, p = nil)
        if(p.nil?)
          @name = name
          @death = true
          @canISteal = true
          @visibleTreasures = Array.new
          @hiddenTreasures = Array.new
          @enemy = nil
          @pendingBadConsequence = nil 
          @level = 1
        else
          @name = p.name
          @death = p.death
          @canISteal = p.canISteal
          @visibleTreasures = p.visibleTreasures
          @hiddenTreasures = p.hiddenTreasures
          @enemy = p.enemy
          @pendingBadConsequence = p.pendingBadConsequence
          @level = p.level
        end
  end
  
  def getVisibleTreasures
    return @visibleTreasures
  end
  
  def getHiddenTreasures
    return @hiddenTreasures
  end
  
  def getOponentLevel(m)
    return m.combatLevel
  end
  
  def bringToLife
    @death = false
  end

  def getCombatLevel
    bonus = 0
    @visibleTreasures.each { |v|
      bonus = bonus + v.bonus
    } 
        
    return @level + bonus
  end
  
  def shouldCovert()
    dice = Dice.getInstance
    number = dice.nextNumber
    if(number == 6) 
      return true      
    else
      return false
    end
  end

  def incrementLevels(l)
    @level = @level + l
  end
  
  def decrementLevels(l) 
    if((@level - l) < 1)
      @level = 1
    else
      @level = @level - l
    end
  end

  def setPendingBadConsequence(b)
    @pendingBadConsequence = b
  end

  def applyPrize(m) 
    nLevels = m.getLevelsGained
    incrementLevels(nLevels)
    nTreasures = m.getTreasuresGained
    if(nTreasures > 0)
      dealer = CardDealer.getInstance
      
      for i in (0..(nTreasures - 1))
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end
  end

  def applyBadConsequence(m) 
    badConsequence = m.bc
    nLevels = badConsequence.someLevels
    decrementLevels(nLevels)    
    pendingBad = badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
    setPendingBadConsequence(pendingBad)
  end

  def canMakeTreasureVisible(t) 
    canMakeVisible = true
        
    contOneHand = 0
    @visibleTreasures.each { |tr| 
      if((tr.treasureKind == t.treasureKind) && (t.treasureKind != TreasureKind::ONEHAND)) 
        canMakeVisible = false
      elsif((t.treasureKind == TreasureKind::BOTHHANDS) && (tr.treasureKind == TreasureKind::ONEHAND)) 
        canMakeVisible = false
      elsif((t.treasureKind == TreasureKind::ONEHAND) && (tr.treasureKind == TreasureKind::BOTHHANDS)) 
        canMakeVisible = false
      end
      if(tr.treasureKind == TreasureKind::ONEHAND) 
        contOneHand = contOneHand + 1
      end
    }
        
    if((t.treasureKind == TreasureKind::ONEHAND) && (contOneHand > 1)) 
      canMakeVisible = false
    end
        
    return canMakeVisible
  end

  def howManyVisibleTreasures(tKind)

    nVisibles = 0;
    
    @visibleTreasures.each { |v|
      if(v.treasureKind == tKind)
        nVisibles = nVisibles + 1
      end
    } 
    
    return nVisibles
  end

  def dieIfNoTreasures
        if(@visibleTreasures.empty? && @hiddenTreasures.empty?)
            @death = true
        end 
  end

  def isDead()
    return @death
  end


  def combat(m) 
    combatResult = CombatResult::LOSE
    myLevel = getCombatLevel
    monsterLevel = getOponentLevel(m)     
        
    if(!(canISteal))
      dice = Dice.getInstance
      number = dice.nextNumber
      if(number < 3)
        enemyLevel = @enemy.getCombatLevel
        monsterLevel = monsterLevel + enemyLevel
      end
    end
        
    if(myLevel > monsterLevel)
      applyPrize(m)
      if(@level >= 10)
        combatResult = CombatResult::WINGAME

      else
        combatResult = CombatResult::WIN
      end
      
    elsif(shouldCovert)
      applyBadConsequence(m)
      combatResult = CombatResult::LOSEANDCONVERT
    else
      applyBadConsequence(m)
      combatResult = CombatResult::LOSE
    end
      
        
    return combatResult
  end

  def makeTreasureVisible(t)
   canI = canMakeTreasureVisible(t)
    if(canI) 
      @visibleTreasures << t
      @hiddenTreasures.delete(t)
    end
  end

  def discardVisibleTreasure(t)
    @visibleTreasures.delete(t)
    if((@pendingBadConsequence != nil))
      if(!(@pendingBadConsequence.isEmpty())) 
        @pendingBadConsequence.substractVisibleTreasure(t)
      end
    end
    dieIfNoTreasures
  end

  def discardHiddenTreasure(t)
    @hiddenTreasures.delete(t)
    if((@pendingBadConsequence != nil))
      if(!(@pendingBadConsequence.isEmpty()))
        @pendingBadConsequence.substractHiddenTreasure(t)
      end      
    end
    dieIfNoTreasures
  end

  def validState
    return (((@pendingBadConsequence == nil) || (@pendingBadConsequence.isEmpty)) && (@hiddenTreasures.length <= 4))
  end

  def initTreasures
    dealer = CardDealer.getInstance
    dice = Dice.getInstance
        
    bringToLife
    treasure = dealer.nextTreasure
    @hiddenTreasures << treasure
    number = dice.nextNumber
        
    if(number > 1) 
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
    end
        
    if(number == 6) 
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
    end
  end

  def stealTreasure
    canI = canISteal()   
    treasure = nil
        
    if(canI)            
      canYou = @enemy.canYouGiveMeATreasure
      if(canYou)
        treasure = @enemy.giveMeATreasure
        @hiddenTreasures << treasure
        haveStolen
      end
    end
    return treasure
  end

  def giveMeATreasure
    pos = rand(@hiddenTreasures.length)
         
    return @hiddenTreasures.at(pos)
  end

  def canYouGiveMeATreasure
    return !(@hiddenTreasures.empty?)
  end

  def haveStolen
    @canISteal = false
  end

  def discardAllTreasures
    visibles = Array.new(@visibleTreasures)
    ocultos = Array.new(@hiddenTreasures)
        
    visibles.each { |t|
      discardVisibleTreasure(t)
    }
    
    ocultos.each { |t|
      discardHiddenTreasure(t)
    }
    
  end
  
  def to_s
    "#{@name}"
  end
   
  private :bringToLife, :incrementLevels, :decrementLevels, :setPendingBadConsequence, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures, :haveStolen
  
  protected :MAXLEVEL, :getCombatLevel, :getOponentLevel, :shouldCovert

end

File: PDOO-Napakalaki/Napakalaki/NapakalakiJava/src/NapakalakiGame/Player.java
package NapakalakiGame;
import GUI.Dice;
import java.util.ArrayList;

public class Player {
    private final int MAXLEVEL;
    private String name;
    private int level;
    private boolean dead;
    private boolean canISteal;
    private CardDealer dealer;
    private Dice dice;
    protected boolean isCultist;
    private CombatResult combatResult;
    protected Player enemy;
    private BadConsequence pendingBadConsequence;
    private ArrayList<Treasure>hiddenTreasures=new ArrayList();
    private ArrayList<Treasure>visibleTreasures=new ArrayList();

    protected Player getEnemy() {
        return enemy;
    }
    
    public Player(String name) {
        this.name = name;
        dead=true;
        canISteal=true;
        MAXLEVEL=10;
        level=1;
        isCultist=false;
    }
    
    public Player(Player a){
        MAXLEVEL=a.MAXLEVEL;
        name=a.name;
        level=a.level;
        dead=a.dead;
        canISteal=a.canISteal;
        dealer=a.dealer;
        dice=a.dice;
        combatResult=a.combatResult;
        enemy=a.enemy;
        pendingBadConsequence=a.pendingBadConsequence;
        hiddenTreasures=a.hiddenTreasures;
        visibleTreasures=a.visibleTreasures;
    }

    public boolean IsCultist() {
        return isCultist;
    }

    
    public String getName() {
        return name;
    }

    private void bringToLife(){
       dead=false;
    }
    
    protected  int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        int aux;
        dice=Dice.getInstance();
        aux=dice.nextNumber();
        
        return aux==6;
    }

    protected int getCombatLevel() {

        int aux=this.level;

        for(Treasure tmp : visibleTreasures){
          aux+=tmp.getBonus();
        }
    return aux;
    }

    private void incrementLevels(int i){
        level+=i;
    }

    private void decrementLevels(int i){
        level-=i;

        if(level<1)
             level=1;
    }

    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence=b;
    }

    private void applyPrize(Monster m){

        int nlevels=m.getLevelsGained();
        this.incrementLevels(nlevels);
        int nTreasures=m.getTreasuresGained();

        if(nTreasures>0){
            CardDealer aux = CardDealer.getInstance();

            for(int i=0;i<nTreasures;i++){
                Treasure treasure = aux.nextTreasure();
                this.hiddenTreasures.add(treasure);
            }

        }
    }

    private void applyBadConsequence(Monster m){

        BadConsequence badconsequence = m.getBadConsequence();

        int nlevels=badconsequence.getLevels();
        this.decrementLevels(nlevels);

        BadConsequence pendingBad = badconsequence.adjustToFitTreasureList(hiddenTreasures, hiddenTreasures);
        this.setPendingBadConsequence(pendingBad);
    }

        public boolean canMakeTreasureVisible(Treasure t) {


          boolean visible = true;
          int contador =0;

          for(int i=0; i<visibleTreasures.size() && visible; i++){
              if(visibleTreasures.get(i).getType() == TreasureKind.BOTHHANDS && t.getType() == TreasureKind.ONEHAND)
                  visible=false;
                  else if(visibleTreasures.get(i).getType() == TreasureKind.ONEHAND && t.getType() == TreasureKind.BOTHHANDS)
                    visible=false;
                      else if(t.getType()!= TreasureKind.ONEHAND && visibleTreasures.get(i).getType() == t.getType())
                        visible=false;
                      if(visibleTreasures.get(i).getType() == TreasureKind.ONEHAND)
                          contador++;
                  }
              if(visible && contador == 2 && t.getType() == TreasureKind.ONEHAND)
                  visible=false;

          return visible;
             }


    private int howManyVisibleTreasures(TreasureKind t){
        int tam=0;

        for(Treasure tmp:visibleTreasures){
            if(tmp.getType()==t)
                tam++;
        }
     return tam;
    }

    private void dieIfNoTreasures(){

        if(this.visibleTreasures.isEmpty()&&this.hiddenTreasures.isEmpty())
        dead=true;
    }

    public boolean isDead() {
        return dead;
    }

    public ArrayList<Treasure> getHiddenTreasures() {
        return hiddenTreasures;
    }

    public ArrayList<Treasure> getVisibleTreasures() {
        return visibleTreasures;
    }

    public CombatResult combat(Monster m){
        CombatResult combat;

        int myLevel=this.getCombatLevel();
        int monsterLevel=getOponentLevel(m);

        if(!this.canISteal()){
            Dice aux= Dice.getInstance();
            int number=aux.nextNumber();

            if(number<3){
                int enemyLevel=this.enemy.getCombatLevel();
                monsterLevel+=enemyLevel;
            }
        }

         if(myLevel>monsterLevel){
             this.applyPrize(m);

             if(this.getLevels()>=this.MAXLEVEL)
                 combat=CombatResult.WINGAME;
             else
                 combat=CombatResult.WIN;

        }else{

             if(this.shouldConvert()){
                 combat=CombatResult.LOSEANDCONVERT;
             
             }else{
                 
                this.applyBadConsequence(m);
                combat=CombatResult.LOSE;
             }
         }

    return combat;
    }

    public void makeTreasureVsible(Treasure t){
        boolean canI=this.canMakeTreasureVisible(t);
        if(canI){
            this.visibleTreasures.add(t);
            this.hiddenTreasures.remove(t);
        }
    }

    public void discardVisibleTreasure(Treasure t){
        this.visibleTreasures.remove(t);

        if(this.pendingBadConsequence!=null && (!this.pendingBadConsequence.isEmpty())){
            this.pendingBadConsequence.substractVisibleTreasure(t);
        }

        this.dieIfNoTreasures();
    }

    public void discardHiddenTreasure(Treasure t){

        this.hiddenTreasures.remove(t);

        if(this.pendingBadConsequence!=null && (!this.pendingBadConsequence.isEmpty())){
            this.pendingBadConsequence.substractHiddenTreasure(t);
        }

        this.dieIfNoTreasures();
    }

    public BadConsequence getPendingBadConsequence() {
        return pendingBadConsequence;
    }

    public boolean validState(){
        return (pendingBadConsequence==null || (pendingBadConsequence.isEmpty()&& (hiddenTreasures.size())<4)  );
    }

    public void initTreasures(){

        int number;
        dealer=CardDealer.getInstance();
        dice=Dice.getInstance();
        this.bringToLife();

        Treasure treasure=dealer.nextTreasure();
        this.hiddenTreasures.add(treasure);

        number=dice.nextNumber();

        if(number==6){
           treasure=dealer.nextTreasure();
           this.hiddenTreasures.add(treasure);

           treasure=dealer.nextTreasure();
           this.hiddenTreasures.add(treasure);
        }

        else if(number>1){
            treasure=dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }
    }

    public int getLevels() {
        return level;
    }

    public Treasure stealTreasure(){
        Treasure devolver=null;
        boolean canI=this.canISteal();

        if(canI){
            boolean canYou = enemy.canYouGiveMeATreasure();

            if(canYou){

                devolver=enemy.giveMeATreasure();
                this.hiddenTreasures.add(devolver);
                this.haveStolen();
            }
        }
        return devolver;
    }

    public void setEnemy(Player enemy) {
        this.enemy = enemy;
    }

    private Treasure giveMeATreasure(){


        int indice = (int)(Math.random() * (this.hiddenTreasures.size()-1));
        Treasure devolver=this.enemy.hiddenTreasures.get(indice);
        this.enemy.hiddenTreasures.remove(indice);
        
        return devolver;
    }

    public boolean canISteal(){return canISteal;}

    private boolean canYouGiveMeATreasure(){

      boolean devolver=( !(hiddenTreasures.isEmpty()) || !(visibleTreasures.isEmpty()) );
    return devolver;
    }

    private void haveStolen(){canISteal=false;}

    public void discardAllTreasures(){

        ArrayList<Treasure>visible_aux=new ArrayList(this.visibleTreasures);
        ArrayList<Treasure>hidden_aux=new ArrayList(this.hiddenTreasures);

        for(Treasure tmp : visible_aux){
            this.discardVisibleTreasure(tmp);
            CardDealer.getInstance().giveTreasureBack(tmp);
        }

        for(Treasure tmp : hidden_aux){
            this.discardHiddenTreasure(tmp);
            CardDealer.getInstance().giveTreasureBack(tmp);
    }
  }
    @Override
    public String toString() {
        return "El nombre del jugador es: " + name + ", su nivel es: " + level;
    }
}
